src/client/util/http.cc

```diff
@@ -180,8 +180,9 @@ bool HTTP::DownloadViaClearnet() {
       }
       // Create response object, send request and receive response
       LOG(trace) << "HTTP: Request " << kovri::core::LogNetMessageToString(request);
-      Response response = client.get(request);
+      Response response = client.head(request);
```

Comment:
While this is a nice idea, this doesn't work as currently implemented. We'll not receive a body to the response when we GET below (i.e., we'll never receive the actual content) and subsequently never actually receive a subscription (thus breaking client functionality).

The issue may be a cpp-netlib'ism. TBD.

Reply:
> We'll not receive a body to the response when we GET below

After running some live tests you recommended, this code works for all reachable reseed servers.

I'm working on another branch to address replacing cpp-netlib with Boost.Beast, but put it on hold to work on alpha milestone issues.

Reply:
>reachable reseed servers.

Unfortunately, there appear to be a handful of reseed servers offline (I haven't heard from the reseed maintainer in a long time).

>this code works

As for "works", well, not really. For reseed, for the ones that are online, notice the unnecessary `Timeout::Request` between null body and a fresh GET (this was *before* your rebase):

```
[2018.06.24 00:37:22.696236] [0x00007ffff7fd7740] [debug]   HTTP: constructor URI https://reseed.onion.im/i2pseeds.su3
[2018.06.24 00:37:22.696914] [0x00007ffff7fd7740] [debug]   HTTP: Download Clearnet with timeout : -
[2018.06.24 00:37:22.696948] [0x00007ffff7fd7740] [trace]   HTTP: Cert exists : "/home/anonimal/.kovri/client/certificates/tls/reseed.onion.im.crt"
[2018.06.24 00:37:22.697646] [0x00007ffff7fd7740] [trace]   HTTP: Request Source : "" Dest : ""
Headers : "User-Agent"  : "Wget/1.11.4" |
Body :
         |

[2018.06.24 00:37:22.698116] [0x00007ffff7fd7740] [trace]   HTTP: Response Source : "" Dest : ""
Headers : "Connection"  : "close" | "Content-Disposition"  : "attachment; filename=i2pseeds.su3" | "Content-Length"  : "75934" | "Content-Type"  : "application/octet-stream" | "Date"  : "Sun, 24 Jun 2018 00:37:23 GMT" | "Server"  : "nginx" | "Strict-Transport-Security"  : "max-age=63072000; includeSubdomains;" | "X-Frame-Options"  : "DENY" | "X-Ratelimit-Limit"  : "4" | "X-Ratelimit-Remaining"  : "3" | "X-Ratelimit-Reset"  : "900" |
Body :
         |

[2018.06.24 00:38:07.698498] [0x00007ffff7fd7740] [debug]   Reseed: fetch successful
```

master:

```
[2018.06.24 00:56:22.866001] [0x00007ffff7fd7740] [debug]   HTTP: constructor URI https://reseed.i2p-projekt.de/i2pseeds.su3
[2018.06.24 00:56:22.866640] [0x00007ffff7fd7740] [debug]   HTTP: Download Clearnet with timeout : -
[2018.06.24 00:56:22.866671] [0x00007ffff7fd7740] [trace]   HTTP: Cert exists : "/home/anonimal/.kovri/client/certificates/tls/reseed.i2p-projekt.de.crt"
[2018.06.24 00:56:22.867352] [0x00007ffff7fd7740] [trace]   HTTP: Request Source : "" Dest : ""
Headers : "User-Agent"  : "Wget/1.11.4" |
Body :
         |

[2018.06.24 00:56:22.867788] [0x00007ffff7fd7740] [trace]   HTTP: Response Source : "" Dest : ""
Headers : "Connection"  : "close" | "Content-Description"  : "File Transfer" | "Content-Disposition"  : "attachment; filename=i2pseeds.su3" | "Content-Length"  : "74497" | "Content-Transfer-Encoding"  : "binary
" | "Content-Type"  : "application/octet-stream" | "Date"  : "Sun, 24 Jun 2018 00:57:34 GMT" | "Server"  : "Apache/2.4.10 (Debian)" |
Body :
         |  49 32 50 73 75 33 00 00  |  00 06 02 00 00 10 00 10  |  00 00 00 00 00 01 20 b9  |  00 00 00 03 00 00 00 00  |
         |  00 00 00 00 00 00 00 00  |  31 35 32 39 32 30 32 34  |  36 38 00 00 00 00 00 00  |  65 63 68 65 6c 6f 6e 40  |
         |  6d 61 69 6c 2e 69 32 70  |  50 4b 03 04 14 00 00 00  |  08 00 38 1d d1 4c 78 45  |  f1 dc b0 02 00 00 d3 02  |
         |  00 00 3b 00 1c 00 72 6f  |  75 74 65 72 49 6e 66 6f  |  2d 78 79 53 59 70 68 67  |  33 35 6f 67 64 42 78 51  |
         |  6e 34 7e 78 42 54 44 4a  |  4b 7e 56 38 31 58 66 51  |  6a 37 68 4e 76 51 70 4d  |  64 7e 38 38 3d 2e 64 61  |
         |  74 55 54 09 00 03 5b bc  |  25 5b ff c6 25 5b 75 78  |  0b 00 01 04 e8 03 00 00  |  04 e8 03 00 00 3b 17 c1  |
         |  f8 fe f5 ff ff 8e f3 d2  |  2b a6 c8 4a 5b 78 3e cd  |  ca 57 d9 5f 77 ed ff ed  |  03 7d 4c 5f 77 73 fa 6f  |
         |  bd 16 db 51 f5 fa 29 c7  |  e4 3d 53 59 8e 5a b9 5e  |  09 96 e0 c8 aa 71 7d 77  |  de ef cc d7 76 ae 12 b9  |
         |  2b 39 b3 f7 46 7c 39 ac  |  f8 f4 58 d8 63 ef 6b ce  |  5b 8b 5d 0f 7d 67 3b 7b  |  f2 ea b4 ad 2c 76 72 dd  |
         |  93 02 57 ac 4c e9 5b 74  |  c2 40 54 22 67 ed 46 2d  |  1d 0d 8d 3b 9a 7f 53 6a  |  4e 5d 6a db fa e5 c3 a3  |
         |  57 87 b7 9c fd 21 bf 2d  |  6e e6 a1 fe ca 63 8f b7  |  19 05 f8 7d 9c b7 8e 33  |  fa b4 97 60 2b cb d1 87  |
         |  ac 96 3a fc a7 13 a4 9b  |  27 b9 bc 2a ea 9e bc ee  |  f2 ee d2 83 d7 56 ea 09  |  4c f8 aa 5c 64 3d 61 fa  |
```

Now, *after* your rebase (timeout still exists):

```
[2018.06.24 01:11:15.507041] [0x00007ffff7fd7740] [debug]   HTTP: constructor URI https://reseed.i2p-projekt.de/i2pseeds.su3
[2018.06.24 01:11:15.508031] [0x00007ffff7fd7740] [debug]   HTTP: Download Clearnet with timeout : -
[2018.06.24 01:11:15.508072] [0x00007ffff7fd7740] [trace]   HTTP: Cert exists : "/home/anonimal/.kovri/client/certificates/tls/reseed.i2p-projekt.de.crt"
[2018.06.24 01:11:15.508750] [0x00007ffff7fd7740] [trace]   HTTP: Request Source : "" Dest : ""
Headers : "User-Agent"  : "Wget/1.11.4" |
Body :
         |

[2018.06.24 01:11:15.509200] [0x00007ffff7fd7740] [trace]   HTTP: Response Source : "" Dest : ""
Headers : "Connection"  : "close" | "Content-Description"  : "File Transfer" | "Content-Disposition"  : "attachment; filename=i2pseeds.su3" | "Content-Length"  : "74497" | "Content-Transfer-Encoding"  : "binary" | "Content-Type"  : "application/octet-stream" | "Date"  : "Sun, 24 Jun 2018 01:12:26 GMT" | "Server"  : "Apache/2.4.10 (Debian)" |
Body :
         |

[2018.06.24 01:11:15.660840] [0x00007ffff7fd7740] [trace]   HTTP: Request Source : "" Dest : ""
Headers : "User-Agent"  : "Wget/1.11.4" |
Body :
         |

[2018.06.24 01:11:15.661049] [0x00007ffff7fd7740] [trace]   HTTP: Response Source : "" Dest : ""
Headers : "Connection"  : "close" | "Content-Description"  : "File Transfer" | "Content-Disposition"  : "attachment; filename=i2pseeds.su3" | "Content-Length"  : "74497" | "Content-Transfer-Encoding"  : "binary" | "Content-Type"  : "application/octet-stream" | "Date"  : "Sun, 24 Jun 2018 01:12:26 GMT" | "Server"  : "Apache/2.4.10 (Debian)" |
Body :
         |  49 32 50 73 75 33 00 00  |  00 06 02 00 00 10 00 10  |  00 00 00 00 00 01 20 b9  |  00 00 00 03 00 00 00 00  |
         |  00 00 00 00 00 00 00 00  |  31 35 32 39 32 30 32 34  |  36 38 00 00 00 00 00 00  |  65 63 68 65 6c 6f 6e 40  |
         |  6d 61 69 6c 2e 69 32 70  |  50 4b 03 04 14 00 00 00  |  08 00 38 1d d1 4c 78 45  |  f1 dc b0 02 00 00 d3 02  |
         |  00 00 3b 00 1c 00 72 6f  |  75 74 65 72 49 6e 66 6f  |  2d 78 79 53 59 70 68 67  |  33 35 6f 67 64 42 78 51  |
         |  6e 34 7e 78 42 54 44 4a  |  4b 7e 56 38 31 58 66 51  |  6a 37 68 4e 76 51 70 4d  |  64 7e 38 38 3d 2e 64 61  |
         |  74 55 54 09 00 03 5b bc  |  25 5b ff c6 25 5b 75 78  |  0b 00 01 04 e8 03 00 00  |  04 e8 03 00 00 3b 17 c1  |

...

[2018.06.24 01:12:00.509835] [0x00007ffff7fd7740] [debug]   Reseed: fetch successful
```

**More importantly**, reseed aside, client functionality is still broken re: subscriptions (have you tried using a client tunnel?). Whether the timeout somehow effects subscription handling, I haven't looked, but I cannot use a client tunnel with this branch.

Have you looked at the cpp-netlib docs yet? I haven't, and yes, I too would rather be dealing with boost.beast right now. Perhaps we should move on that sooner than later.

Reply:
> Have you tried implementing a new, short, timeout between the head and get request?

Tried a short timeout before we started working on fixes, but only with reseed. I'll try again to see if that resolves the issue with client tunnels as well.

> Must you tear down the stream before submitting a get request or can you use the existing one?

I have no idea, and can look into the cpp-netlib docs/code to find an answer.

The simplest solution I saw was using the `head/get()` methods, which basically just construct an object with one flag switched for retrieving the response body.

If it turns out the body is being downloaded anyway, I'll discard this PR in favor of more overarching Boost.Beast refactors.

> I too would rather be dealing with boost.beast right now. Perhaps we should move on that sooner than later.

If this is a priority, I can spend some hours refactoring with Boost.Beast, and the cpp-netlib URI library.

Also been looking at how I can contribute to the WIP Boost.Beast URI library, but the hybrid approach seems more attainable atm. I need to get more familiar with the URI RFC, and the advanced techniques used in the WIP URI library.